File: algorithms-data-structures/ts/src/data-structures/tree/bst/tree.ts
interface BinarySearchTreeAble {
  insert(value: number): TreeNode | null;
  insertIter(value: number): TreeNode | null;
  find(value: number): TreeNode | null;
  remove(value: number): TreeNode | null;
  contains(value: number): boolean;
  size(): number;
  depth(): number;
  min(): number;
  max(): number;
  isEmpty(): boolean;
  toArray(): number[];
  toString(): string;
  bfs(): number[];
  dfs(order: "PRE" | "POST" | "IN"): number[];
}

class TreeNode {
  value: number;
  left: TreeNode | null;
  right: TreeNode | null;

  constructor(value: number) {
    this.value = value;
    this.left = null;
    this.right = null;
  }
}

export class BinarySearchTree implements BinarySearchTreeAble {
  root: TreeNode | null;

  constructor() {
    this.root = null;
  }

  insertIter(value: number): TreeNode | null {
    if (this.root === null) {
      this.root = new TreeNode(value);
      return this.root;
    }
    let node = this.root;
    while (node !== null) {
      if (value < node.value) {
        if (node.left === null) {
          node.left = new TreeNode(value);
          return node.left;
        }
        node = node.left;
      } else {
        if (node.right === null) {
          node.right = new TreeNode(value);
          return node.right;
        }
        node = node.right;
      }
    }
    return null;
  }

  insert(value: number): TreeNode | null {
    if (this.root === null) {
      this.root = new TreeNode(value);
      return this.root;
    }
    return this.#insertHelper(this.root, value);
  }

  contains(value: number): boolean {
    return false;
  }
  max(): number {
    if (this.root === null) return -1;
    let current = this.root;
    while (current.right !== null) {
      current = current.right;
    }
    return current.value;
  }
  bfs(): number[] {
    if (this.root === null) return [];
    let result: number[] = [];
    let queue: TreeNode[] = []; // TODO make a real Q data structure
    queue.push(this.root);
    while (queue.length > 0) {
      let node = queue.shift();
      if (node) {
        result.push(node?.value);
        if (node.left) {
          queue.push(node.left);
        }
        if (node.right) {
          queue.push(node.right);
        }
      }
    }
    return result;
  }

  find(value: number): TreeNode | null {
    if (this.root === null) return null;
    return this.#findHelper(this.root, value);
  }

  remove(value: number): TreeNode | null {
    throw new Error("Method not implemented.");
  }
  size(): number {
    if (this.root === null) return 0;
    return this.#sizeHelper(this.root);
  }

  depth(): number {
    throw new Error("Method not implemented.");
  }
  min(): number {
    if (this.root === null) return -1;
    let current = this.root;
    while (current.left !== null) {
      current = current.left;
    }
    return current.value;
  }
  isEmpty(): boolean {
    return this.root === null;
  }
  toArray(): number[] {
    throw new Error("Method not implemented.");
  }
  toString(): string {
    throw new Error("Method not implemented.");
  }
  dfs(order: "PRE" | "POST" | "IN" = "PRE"): number[] {
    let result: number[] = [];
    switch (order) {
      case "PRE":
        this.#preOrder(this.root, result);
        break;
      case "POST":
        this.#postOrder(this.root, result);
        break;
      case "IN":
        this.#inOrder(this.root, result);
        break;
      default:
        break;
    }
    return result;
  }
  #inOrder(node: TreeNode | null, result: number[]) {
    if (node === null) return;
    this.#inOrder(node.left, result);
    result.push(node.value);
    this.#inOrder(node.right, result);
  }

  #preOrder(node: TreeNode | null, result: number[]) {
    if (node === null) return;
    result.push(node.value);
    this.#preOrder(node.left, result);
    this.#preOrder(node.right, result);
  }

  #postOrder(node: TreeNode | null, result: number[]) {
    if (node === null) return;
    this.#postOrder(node.left, result);
    this.#postOrder(node.right, result);
    result.push(node.value);
  }

  #sizeHelper(node: TreeNode | null): number {
    if (node === null) return 0;
    return 1 + this.#sizeHelper(node.left) + this.#sizeHelper(node.right);
  }
  #insertHelper(node: TreeNode | null, value: number) {
    if (node === null) {
      node = new TreeNode(value);
      return node;
    }
    if (value < node.value) {
      node.left = this.#insertHelper(node.left, value);
    } else {
      node.right = this.#insertHelper(node.right, value);
    }
    return node;
  }
  #findHelper(node: TreeNode | null, value: number): TreeNode | null {
    if (node === null) return null;
    if (node.value === value) return node;
    if (value < node.value) {
      return this.#findHelper(node.left, value);
    }
    return this.#findHelper(node.right, value);
  }
}

File: algorithms-data-structures/go/src/data-structures/trees/bst/tree.go
package bst

import (
	ll "github.com/bahlo/generic-list-go"
)

type Node struct {
	Value int
	Left  *Node
	Right *Node
}

type Tree interface {
	Insert(value int)
	Remove(value int)
	Size() int
	Find(value int) *Node
	Contains(value int) bool
	BFS() []int
	DFS(order string) []int // in, pre, post

}

func NewBst() Tree {
	return &Bst{}
}

type Bst struct {
	root *Node
}

func (t *Bst) Insert(value int) {
	if t.root == nil {
		t.root = &Node{Value: value}
		return
	}
	insert(t.root, value)
}

func insert(node *Node, value int) *Node {
	if node == nil {
		return &Node{Value: value}
	}
	if value < node.Value {
		node.Left = insert(node.Left, value)
	} else if value > node.Value {
		node.Right = insert(node.Right, value)
	}
	return node
}

func (t *Bst) Remove(value int) {
	if t.root == nil {
		return
	}

}

func (t *Bst) Size() int {
	if t.root == nil {
		return 0
	}
	return size(t.root)
}
func size(node *Node) int {
	if node == nil {
		return 0
	}
	return 1 + size(node.Left) + size(node.Right)
}

func (t *Bst) Find(value int) *Node {
	if t.root == nil {
		return nil
	}
	return find(t.root, value)
}

func find(node *Node, value int) *Node {
	if node == nil {
		return nil
	}
	if value == node.Value {
		return node
	}
	find(node.Left, value)
	find(node.Right, value)
	return nil
}

func (t *Bst) Contains(value int) bool {
	return t.Find(value) != nil
}

func (t *Bst) BFS() []int {
	var result []int
	queue := ll.New[*Node]()
	queue.PushBack(t.root)
	for queue.Len() > 0 {
		element := queue.Front()
		node := element.Value
		queue.Remove(element)
		result = append(result, node.Value)
		if node.Left != nil {
			queue.PushBack(node.Left)
		}
		if node.Right != nil {
			queue.PushBack(node.Right)
		}
	}
	return result
}

func (t *Bst) DFS(order string) []int {
	var result []int
	if t.root == nil {
		return result
	}
	switch order {
	case "IN":
		inOrder(t.root, &result)
	case "PRE":
		preOrder(t.root, &result)
	case "POST":
		postOrder(t.root, &result)
	}
	return result
}
func inOrder(node *Node, result *[]int) {
	if node == nil {
		return
	}
	inOrder(node.Left, result)
	*result = append(*result, node.Value)
	inOrder(node.Right, result)
}
func preOrder(node *Node, result *[]int) {
	if node == nil {
		return
	}
	*result = append(*result, node.Value)
	preOrder(node.Left, result)
	preOrder(node.Right, result)
}
func postOrder(node *Node, result *[]int) {
	if node == nil {
		return
	}
	postOrder(node.Left, result)
	postOrder(node.Right, result)
	*result = append(*result, node.Value)
}
